### ISSUE TITLE
Allow PSXACT to be used as a library

### ISSUE THREAD

**Issue by adam-becker** [adam-becker/psxact#13]
Currently, there is no freedom in regards to front-ends. Users are locked into using the CLI and SDL to run PSXACT.

In hopes of attracting more developers, being able to consume PSXACT as a library would make it easier to develop front-ends for various platforms.

**Comment by wiired24** [comment]
I definitley would want this. You do this and I'll donate to you lol

**Comment by adam-becker** [comment]
Which BIOS is being used feels like a configuration thing more than a you-have-to-do-this-every-single-time thing.

**Comment by wiired24** [comment]
You're Right. Yeah the user would just set it once and only once. I should have clarified that.

**Comment by wiired24** [comment]
Here is a short video showing the current state of the U.I. for psx-box. I think it's coming along nicely 
https://youtu.be/q6fP32AjIg8

**Comment by wiired24** [comment]
So this is actually the latest rendition of the U.I.. As I mentioned previously I had began writing the U.I in C++/XAML but I've since switched over to C#/XAML for the User Interface. I was wanting to get your opinion on what you think. Personally I think this design looks more fiiting for a console .  https://camo.githubusercontent.com/8d1c5421998936eba4a4ff31b63f7c7ceced980e/687474703a2f2f692e696d6775722e636f6d2f614a706943686b2e706e67

I've already Implemented the License/About buttons ("Just used Speech Synthesis to give the user relevant details e.g. for license when they click on the button they would hear "PSX-BOX is licensed under the GPL V3".  All I have left to do is just import/export which for that I'll just Windows::Storage::FileIO Methods to get and send the files to and from the system. Lastly, I'll have to implement some basic config stuff in settings. The U.I could possibly be finished within a few weeks.

**Comment by adam-becker** [comment]
Looks good.

Would building this project automatically ease development of your tool with this proposed library strategy? I'm looking into setting up automated builds for various reasons, your use case being one of them.

Unix builds would use TravisCI. I'm unsure of the Windows equivalent for C/C++ projects.. Is it Appveyor?

**Comment by wiired24** [comment]
From what I understand since I am writing the user interface in C#/XAML and
the emulator core itself is written in C++ I'll need to reference it in the
C# U.I. via a Windows Runtime Component. I don't know much about Windows
Run Time Components as I have never had to do one before but I'm going to
be trying to get some useful information from Stack Overflow. Also I wanted
to point out that I know in the original codebase you used fopen to open a
file etc.  With regards to psx-box however the U.I uses an API Class called
FilePicker which handles opening files etc. This wouldn't conflict with the
emulator would it? And I'm unsure I've never heard of or used appveyor tbh.

**Comment by adam-becker** [comment]
I always figured your GUI was being programmed in C++ for some reason. The way you'll interact with the emulator is with P/Invoke then, so you'll be wanting a `.dll` file.

```csharp
class NativeMethods {
  [DllImport("psxact", EntryPoint = "init")]
  public static extern IntPtr Create();

  [DllImport("psxact", EntryPoint = "run_for_one_frame")]
  public static extern void RunForOneFrame(IntPtr handle);
}

...

var handle = NativeMethods.Create();

while (true) {
  NativeMethods.RunForOneFrame(handle);
}

```

**Comment by wiired24** [comment]
Originally The U.I. was being written in C++/XAML but i switched it to
C#/XAML as there was hardly any documentation covering C++/XAML for UWP
U.I.  Also, Thanks for the information. Also another question which crossed
my mind earlier say the user selects a game for instance "crash bandicoot"
but it could be any game. The file gets selected, it then gets stored in
memory, when main() gets called for the emulator to initialize and start up
how will it know which game has been passed in? I know this may sound
confusing but here take a look at a short video to see how I have
implemented the U.I. Maybe that will give you a better Idea of what I'm
trying to ask. I've pretty much got everything implemented except for
getting the psxact.dll core itself

https://youtu.be/_npU-fmHUQg

**Comment by adam-becker** [comment]
Since the emulator is being rewritten as a library, you'll have an obvious way of passing in the game and bios file names. Probably this will manifest as parameters to the `init` function.

I'm still in the planning stages of how this should happen, I've pretty much settled on the following basic outline:

- The emulator is just that, a library for creating/running a simulation of a game system (PSX in this case). It will not have any dependency on any 3rd party library because...
- The front-end handles drawing the screen, rendering audio and getting input states.

This satisfies the goal of creating the emulator as a library, and drastically simplifies the emulator itself. However this creates more of a burden on the front-end developer. So at the moment I am thinking of ways to make the change without dropping a lot of responsibility in your lap :smile:

Perhaps making my own basic CLI front-end that references PSXACT as a library will shed some light on the pitfalls with this approach.

**Comment by wiired24** [comment]
Yeah that sounds great to me. I tried inquring on StackOverflow about how
to take the Emulator (psxact) and import the C++ code into the C# U.I.
through a Windows Run time componenet. But it was closed because it was
"Too broad". So i didn't get any answers there (Go Figure). I know you said
I could use p_invoke but isn't that just for global C++ functions? Or will
pinvoke work just fine with the library that you're going to be making?
Because If i can just use p_invoke to call the emulator I'd rather do it
that way vs creating a run time component and linking this with that etc etc

**Comment by adam-becker** [comment]
This is where we get into an area I'm not well versed in. XBOX One may run under partial trust settings, but I'd hope it allows P/Invoking libraries.

The public interface for the library will be under `export "C" { ... }` so will have a basic C function export signature, allowing you to P/Invoke easily.

**Comment by wiired24** [comment]
Even if p_invoke isn't allowed with UWP I'll just have to use your library as a Dll and reference it in the Front End with a Windows RunTime Component. Either way we're good.

**Comment by wiired24** [comment]
P invoke actually seems to be working just fine with UWP.   Roughly speaking how long do you think it could take to write the library? Also correct me if I am wrong but I noticed that in this case p_invoke begins it's entry point at "init". I'm assuming that's a function call in psxact that initalizes the Emulator?

Also I have a ButtonClick Method which uses FilePicker to import a game file from External/Internal HDD

            // Uses File Picker to Import a Game
            var picker = new Windows.Storage.Pickers.FileOpenPicker();
            picker.ViewMode = Windows.Storage.Pickers.PickerViewMode.Thumbnail;
            picker.SuggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.DocumentsLibrary;
            picker.FileTypeFilter.Add(".bin");
            Windows.Storage.StorageFile file = await picker.PickSingleFileAsync();

            if (file != null)
            {
       
          var handle = NativeMethods.Create();

          while (true) {
          NativeMethods.RunForOneFrame(handle);
          }

I'm assuming that var handle while (True) would go into my if (file != null) Logic?

**Comment by adam-becker** [comment]
We'll have to see how this works, because at the moment PSXACT can't even load game files (CDROM is only emulated enough to tell the BIOS no game is in the tray). I'm hoping that you'd just be able to pass in the file names, probably as parameters to `init`:

```c++
void *init(const char *bios_file_name, const char *game_file_name);
```

The `init` function will create a new `state_t` structure, and return a pointer to it. You'll then have to pass that state pointer back to the library in the `run_for_one_frame` function. `run_for_one_frame` will likely also need controller data that the front end will provide:

```c++
enum button_state_t {
  BUTTON_STATE_PRESSED,
  BUTTON_STATE_RELEASED
};

struct input_state_t {
  button_state_t triangle;
  button_state_t square;
  button_state_t circle;
  button_state_t cross;
  // etc ...
};
```

I'll also have to provide a way to get the frame buffer data, but this I'm also unsure of how I want to go about this, since a game can draw anywhere in the frame buffer it wants so giving a simple pointer isn't going to be wise.

Perhaps I can handle this with a struct describing the current video state:

```c++
struct video_state_t {
  unsigned width;
  unsigned height;

  struct {
    unsigned stride;
    uint16 *frame_buffer;
  } pixel_data;
};
```

I'm pretty certain (I'd have to check again) that all data written to the frame buffer by draw operations is 16-bit BGR. You'd then get the video state as a return value from `run_for_one_frame`:

```c++
video_state_t run_for_one_frame(void *state, input_state_t *input1, input_state, input_state_t *input2);
```

I'll more than likely expose audio data in a similar way, but I haven't even started emulating the SPU registers, so I can't say for certain how the API surface will look there.

**Comment by wiired24** [comment]
Sounds like there is a ton of work still left to do. Between implementing Audio, SPU Registers, improving the emulator to the point where we can get some commercial games booting. I'm more of the front end guy you are more of the engineer. I only understand the fundamental basics when it comes to how an emulator works i.e. Dissassembler (to get the opcodes), Operation codes, registers, cpu cycles etc. However with both of us working in concert on this we can have a really nice PS1 Emulator. I also plan on releasing psx-box for Windows 10 PC as well as Xbox One. Could possibly run on Linux with mono but i'm unsure. The only remaining piece of the puzzle for psx-box is just the core emulator library itself. Everything else is mostly finished.

**Comment by adam-becker** [comment]
Honestly, CDROM emulation is what is holding PSXACT back the most. I am very new to disc based gaming consoles, and have very little understanding of how a CDROM drive works in practice. This weekend I'll be flowcharting the FSM and hopefully be able to write some more code for that module.

Audio is a lesser concern, since most games can be played just fine without audio. The GPU is probably the most fun part of the project thus far, allowing me to use math I never thought I'd use.

**Comment by wiired24** [comment]
Yeah but I'm confident you will figure it out though. I don't know how much it might help but you could always study other open source psx emulators like pcsx-reloaded and see how they did the cd-rom code. It may or may not explain the "why" but at least you'd get some practical insight. Audio would be nice but my main concern is compatibility and having a good number of games running before tackling audio. That can always be done later.

**Comment by wiired24** [comment]
How is the FSM Coming along any progress on that front?

**Comment by adam-becker** [comment]
So far this weekend I've just been nursing my dog, he twisted his wrist so I have to watch him pretty much 24/7. Haven't made as much progress as I'd like, but I was able to fix some issues in the CDROM that were low-hanging fruit.

I looked at the `mednafen` implementation, and the code isn't very easy to follow. One of the reasons I don't like reading other people's source code 😄 I really should invest some money into getting a real PSX that I can just write tests for to guide the emulator's development, not sure how doable that is at the moment.

**Comment by wiired24** [comment]
Yeah it's hard sometimes to read other peoples code. I always try my best to document what I'm doing and why I'm doing it when I write code but sometimes it can be hard to always stick to that. Sorry to hear about your dog. Hope everything is going fine for Him/Her. 

I've been busy this weekend working on porting Chip8 to Xbox One, I actually just finished it today.
https://gbatemp.net/threads/how-to-sideload-chip8-emulator-on-xbox-one.470025/#post-7295485

As far as psx-box is concerned it's pretty much finished. I'll Just need to implement PSXACT as a library when the emulator  gets to a more useable state.

**Comment by adam-becker** [comment]
I'm starting an IRC channel for this project, I'll update the `README` with details.

**Comment by wiired24** [comment]
Alright Cool  We can just continue communication on there.

**Comment by zminhquanz** [comment]
@wiired24 that's look good on UWP ( Windows 10 Mobile ) , but on Windows , you can using QT or Wxwidget

**Comment by SigmaVirus** [comment]
@adam-becker 
Front-end? Various platforms? Why not team up with RetroArch and make a psxact Libretro core?

**Comment by adam-becker** [comment]
@SigmaVirus I have no experience with Libretro, so I wouldn't know where to begin. I'm not against the idea though.

**Comment by SigmaVirus** [comment]
@adam-becker 
I made a small post in the [Core and Feature Requests thread](https://forums.libretro.com/t/core-and-feature-requests/5895/256) regarding psxact in the Libretro forums hoping to get some attention, catching some interest about the idea of making a Libretro core out of it.

We'll see if it it worked or not, I know I'd support it because mednafen PSX needs competition with fresh thinking.

You say you have no experience with Libretro, does that mean you haven't tried RetroArch?
If you haven't I strongly recommend you to do because that will give you a nice opportunity to see what RetroArch is capable of and is all about.

You'll find what you need [here](http://buildbot.libretro.com/nightly/)
